### ISSUE TITLE
torch.stack does not yet support named tensors 

### ISSUE THREAD

**Issue by ricpruss** [pytorch/pytorch#29010]
## 🚀 Feature
tensor.stack outputs named tensors.

## Motivation

Stack is one of those times when you are mushing together a bunch of tensors and getting the end result to match what you intended can be surprisingly tedious and finicky. Named tensors is a beautiful idea that provides names for each dimension.

## Pitch

The input tensors should be named and the names should match in all the dimensions.
e..g.
```
test1 = torch.from_numpy(np.array([[4,1,5,3],[0,6,-1,3.]])).refine_names('H', 'W')
test2 = torch.from_numpy(np.array([[1,1,2,2],[1,1,3,8.]])).refine_names('H', 'W')
newstack = torch.stack((test1,test2), dim=0,  names=('B', 'H', 'W'))
```
'B' is the new dimension you are stacking, other dimensions should match by name.
 
## Alternatives

The alternative is the current implementation where you get a clear error.
`RuntimeError: stack is not yet supported with named tensors. Please drop names via `tensor = tensor.rename(None)`, call the op with an unnamed tensor, and set names on the result of the operation.`

## Additional context

Write up about named tensors. [Named Tensors for Deep Learning](http://nlp.seas.harvard.edu/NamedTensor2)

**Comment by jonasteuwen** [comment]
Agree, also came up in my use case: collate_fn uses stack. See https://discuss.pytorch.org/t/collating-named-tensors/78650

**Comment by jonasteuwen** [comment]
`torch.stack` supporting named tensors would be very useful as this is also used in e.g. stacking them up in batches. I can create a pull request implementing it, but there is currently no other similar operation implementing adding a new dimension, so unclear what it's api would be. `torch.stack([...], names=[...])` where names add one new dimension and the index is inferred from this? 
Or `torch.stack([...], idx)` where the new dimension is just named 'None'? The latter would likely be the simplest to do, and the name can be refined.

cc @zou3519
